pkg/noderesourcetopology/resource_allocation_score.go

```diff
@@ -0,0 +1,155 @@
+package noderesourcetopology
+
+import (
+	"context"
+	"fmt"
+
+	topologyv1alpha1 "github.com/k8stopologyawareschedwg/noderesourcetopology-api/pkg/apis/topology/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+	apiconfig "sigs.k8s.io/scheduler-plugins/pkg/apis/config"
+)
+
+type ScoreStrategy string
+
+const (
+	// ScorePluginName is the name of the plugin used in the plugin registry and configurations.
+	ScorePluginName = "NodeResourceTopologyResourceAllocationScore"
+
+	// LeastAllocatable strategy favors node with the least amount of available resource
+	LeastAllocatable ScoreStrategy = "least-allocatable"
+
+	// BalancedAllocation strategy favors nodes with balanced resource usage rate
+	BalancedAllocation ScoreStrategy = "balanced-allocation"
+
+	// MostAllocatable strategy favors node with the most amount of available resource
+	MostAllocatable ScoreStrategy = "most-allocatable"
+
+)
+
+// resourceAllocationScorer contains information to calculate resource allocation score.
+type resourceAllocationScorer struct {
+	scoreStrategy       scoreStrategy
+	resourceToWeightMap resourceToWeightMap
+	data                commonPluginsData
+}
+
+var _ framework.ScorePlugin = &resourceAllocationScorer{}
+
+func (r *resourceAllocationScorer) Name() string {
+	return ScorePluginName
+}
+
+func (r *resourceAllocationScorer) Score(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	// calculates the fraction of requested to capacity per each numa-node.
+	// return the numa-node with the minimal score as the node's total score
+	return r.score(pod, nodeName)
+}
+
+func (r *resourceAllocationScorer) ScoreExtensions() framework.ScoreExtensions {
+	return nil
+}
+
+// score will use `scoreStrategy` function to calculate the score.
+func (r *resourceAllocationScorer) score(pod *v1.Pod, nodeName string) (int64, *framework.Status) {
+	klog.V(5).Infof("Call score for node %v", nodeName)
+	nodeTopology := findNodeTopology(nodeName, &r.data)
+
+	if nodeTopology == nil {
+		return 0, nil
+	}
+
+	klog.V(5).Infof("nodeTopology: %v", nodeTopology)
+	calculateResourceAllocatableRequest(pod, nodeTopology)
+	resources, allocatablePerNUMA := calculateResourceAllocatableRequest(pod, nodeTopology)
+	return scoreHelper(resources, allocatablePerNUMA, r.scoreStrategy, r.resourceToWeightMap)
+}
+
+// scoreHelper will iterate over all NUMA zones of the node and invoke the scoreStrategy func for every zone.
+// it will return the minimal score of all the calculated NUMA's score, in order to avoid edge cases.
+func scoreHelper(requested v1.ResourceList, numaList NUMANodeList, score scoreStrategy, resourceToWeightMap resourceToWeightMap ) (int64, *framework.Status) {
```

Comment:
scoreHelper is too general, it seems it is find or calculate Min (or omit min) ScoreForEachNUMANode